File: simulate-branches.py
"""
Prism branch simulator — Python

Sends events from 4 fake branches without you needing to git-switch.
Each branch gets a different number of calls to generate realistic spread.

After running, your Prism dashboard → GitHub tab should show:
  main          — 4 calls   (highest cost, widest bar)
  feat/auth     — 3 calls
  feat/search   — 2 calls
  fix/typo      — 1 call    (lowest cost)

Run:
  python simulate-branches.py
"""

import os
import sys
import time
from dotenv import load_dotenv

load_dotenv()

if not os.environ.get("OPENAI_API_KEY") or not os.environ.get("PRISM_API_KEY"):
    print("❌  Set OPENAI_API_KEY and PRISM_API_KEY in .env first.")
    sys.exit(1)

BRANCHES = [
    {
        "name":   "main",
        "commit": "a1b2c3d",
        "calls": [
            "Summarise the CAP theorem in two sentences.",
            "What are the SOLID principles? One line each.",
            "Explain eventual consistency with a real-world analogy.",
            "What is a bloom filter and when would you use one?",
        ],
    },
    {
        "name":   "feat/auth",
        "commit": "e4f5g6h",
        "calls": [
            "Write a short JWT validation function in Python.",
            "What is the difference between OAuth 2.0 and OpenID Connect?",
            "Explain PKCE in one paragraph.",
        ],
    },
    {
        "name":   "feat/search",
        "commit": "i7j8k9l",
        "calls": [
            "What is BM25 and how does it differ from TF-IDF?",
            "Describe how Elasticsearch sharding works in two sentences.",
        ],
    },
    {
        "name":   "fix/typo",
        "commit": "m1n2o3p",
        "calls": [
            "Correct this sentence: 'Their going to the store'.",
        ],
    },
]

print("─" * 60)
print("  Prism Branch Simulator (Python)")
print("─" * 60)
print(f"  Simulating {len(BRANCHES)} branches × their calls")
print("─" * 60)
print()

grand_total = 0.0

for branch in BRANCHES:
    # Override env vars so the SDK picks up this branch's context
    os.environ["GITHUB_REF_NAME"] = branch["name"]
    os.environ["GITHUB_SHA"]      = branch["commit"]

    # Import after setting env so _detect_git_context() picks up the override
    # Re-instantiate per branch so git context is re-read each time
    from prism import OpenAI
    client = OpenAI()

    calls      = branch["calls"]
    branch_cost = 0.0

    print(f"📌 Branch: {branch['name']}  ({len(calls)} calls)")

    for i, prompt in enumerate(calls):
        print(f"   [{i+1}/{len(calls)}] {prompt[:50]}… ", end="", flush=True)

        t0  = time.monotonic()
        res = client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[{"role": "user", "content": prompt}],
            max_tokens=150,
        )
        ms = int((time.monotonic() - t0) * 1000)

        usage    = res.usage
        in_tok   = getattr(usage, "prompt_tokens", 0) or 0
        out_tok  = getattr(usage, "completion_tokens", 0) or 0
        cost     = ((in_tok * 0.15) + (out_tok * 0.60)) / 1_000_000
        branch_cost += cost
        grand_total  += cost

        print(f"done  ({ms}ms · ${cost:.5f})")

    print(f"   └─ Branch total: ${branch_cost:.5f}")
    print()

print("─" * 60)
print(f"  Total cost across all branches: ${grand_total:.5f}")
print("─" * 60)
print()

# Wait for daemon threads to flush before exit
time.sleep(2)

app_url = os.environ.get("PRISM_APP_URL", "https://prism-dip-dey-s-projects.vercel.app")
print("  ✅ All events sent. View in Prism:")
print()
print(f"  {app_url}/dashboard")
print("  → Projects → your project → GitHub tab")
print("  → Branch Activity shows 4 branches with relative cost bars")
print("  → Click any branch in the timeline to see detail")
print()

File: simulate-branches.mjs
/**
 * Prism branch simulator
 *
 * Sends events from 4 fake branches without you needing to git-switch.
 * Each branch gets a different number of calls to generate realistic spread.
 *
 * After running, your Prism dashboard → GitHub tab should show:
 *   main          — 4 calls   (highest cost, widest bar)
 *   feat/auth     — 3 calls
 *   feat/search   — 2 calls
 *   fix/typo      — 1 call    (lowest cost)
 *
 * Run:
 *   node --env-file=.env simulate-branches.mjs
 */

import { OpenAI } from "@prism-llm-labs/sdk";
import { execSync } from "child_process";

if (!process.env.OPENAI_API_KEY || !process.env.PRISM_API_KEY) {
  console.error("❌  Set OPENAI_API_KEY and PRISM_API_KEY in .env first.");
  process.exit(1);
}

const BRANCHES = [
  {
    name:   "main",
    commit: "a1b2c3d",
    calls: [
      "Summarise the CAP theorem in two sentences.",
      "What are the SOLID principles? One line each.",
      "Explain eventual consistency with a real-world analogy.",
      "What is a bloom filter and when would you use one?",
    ],
  },
  {
    name:   "feat/auth",
    commit: "e4f5g6h",
    calls: [
      "Write a short JWT validation function in TypeScript.",
      "What is the difference between OAuth 2.0 and OpenID Connect?",
      "Explain PKCE in one paragraph.",
    ],
  },
  {
    name:   "feat/search",
    commit: "i7j8k9l",
    calls: [
      "What is BM25 and how does it differ from TF-IDF?",
      "Describe how Elasticsearch sharding works in two sentences.",
    ],
  },
  {
    name:   "fix/typo",
    commit: "m1n2o3p",
    calls: [
      "Correct this sentence: 'Their going to the store'.",
    ],
  },
];

console.log("─".repeat(60));
console.log("  Prism Branch Simulator");
console.log("─".repeat(60));
console.log(`  Simulating ${BRANCHES.length} branches × their calls`);
console.log("─".repeat(60));
console.log();

let grandTotal = 0;

for (const branch of BRANCHES) {
  // Override env vars so the SDK picks up this branch's context
  process.env.GITHUB_REF_NAME = branch.name;
  process.env.GITHUB_SHA      = branch.commit;

  // Create a fresh client per branch so git context is re-read
  const client = new OpenAI();

  console.log(`📌 Branch: ${branch.name}  (${branch.calls.length} calls)`);

  let branchCost = 0;

  for (const [i, prompt] of branch.calls.entries()) {
    process.stdout.write(`   [${i + 1}/${branch.calls.length}] ${prompt.slice(0, 50)}… `);

    const t0  = Date.now();
    const res = await client.chat.completions.create({
      model:      "gpt-4o-mini",
      messages:   [{ role: "user", content: prompt }],
      max_tokens: 150,
    });
    const ms = Date.now() - t0;

    const usage   = res.usage;
    const inTok   = usage?.prompt_tokens     ?? 0;
    const outTok  = usage?.completion_tokens ?? 0;
    const cost    = ((inTok * 0.15) + (outTok * 0.60)) / 1_000_000;
    branchCost   += cost;
    grandTotal   += cost;

    console.log(`done  (${ms}ms · $${cost.toFixed(5)})`);
  }

  console.log(`   └─ Branch total: $${branchCost.toFixed(5)}`);
  console.log();
}

console.log("─".repeat(60));
console.log(`  Total cost across all branches: $${grandTotal.toFixed(5)}`);
console.log("─".repeat(60));
console.log();

// Small delay to let background fetch threads complete before exit
await new Promise(r => setTimeout(r, 2000));

const appUrl = process.env.PRISM_APP_URL ?? "https://prism-dip-dey-s-projects.vercel.app";
console.log("  ✅ All events sent. View in Prism:");
console.log();
console.log(`  ${appUrl}/dashboard`);
console.log("  → Projects → your project → GitHub tab");
console.log("  → Branch Activity shows 4 branches with relative cost bars");
console.log("  → Click any branch in the timeline to see detail");
console.log();
